Shorten UpdateNodeLabelsAndRecordEvents function in nmc_reconciler.go

TomerNewman opened the pull request:
## Summary
Today, the UpdateNodeLabelsAndRecordEvents function has around 100 lines of code. This commit aims to enhance its readability and maintainability by using helper functions, thereby reducing its total length. Additionally, adding unit tests that covers the new helper functions.

TomerNewman commented:
/assign @yevgeny-shnaidman

yevgeny-shnaidman commented:
/ok-to-test

yevgeny-shnaidman commented:
It looks good to me , but i have another suggestion:
most of the code of UpdateNodeLabelsAndRecordEvents function deals with preparing the labels for the node. What if we define another helper interface, that will handle that preparation? This way we can unit-test the UpdateNodeLabelsAndRecordEvents properly and in addition test the labels preparation thoroughly also, since it will be in a different interface. @ybettan , @TomerNewman , what do you think? Or is it an overkill?

TomerNewman commented:
> It looks good to me , but i have another suggestion: most of the code of UpdateNodeLabelsAndRecordEvents function deals with preparing the labels for the node. What if we define another helper interface, that will handle that preparation? This way we can unit-test the UpdateNodeLabelsAndRecordEvents properly and in addition test the labels preparation thoroughly also, since it will be in a different interface. @ybettan , @TomerNewman , what do you think? Or is it an overkill?

so it should be in a different package? maybe in `internal->utils->kmmlabels.go?`

yevgeny-shnaidman commented:
> > It looks good to me , but i have another suggestion: most of the code of UpdateNodeLabelsAndRecordEvents function deals with preparing the labels for the node. What if we define another helper interface, that will handle that preparation? This way we can unit-test the UpdateNodeLabelsAndRecordEvents properly and in addition test the labels preparation thoroughly also, since it will be in a different interface. @ybettan , @TomerNewman , what do you think? Or is it an overkill?
> 
> so it should be in a different package? maybe in `internal->utils->kmmlabels.go?`

No, it can be in the same package, for example like PodManager interface

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -565,94 +565,69 @@ func (h *nmcReconcilerHelperImpl) UpdateNodeLabelsAndRecordEvents(ctx context.Co
-			deprecatedNodeModuleReadyLabels.Insert(label)
-		}
-	}
+	getKernelModuleReadyLabels(&nodeModuleReadyLabels, &deprecatedNodeModuleReadyLabels, node)
```
A `get*` function should return something. In your case, looks like a `set*` function will be a better fit.

BUT, I would keep it a `get*` function that return a new `Set` and then intersect the returned value with `nodeModuleReadyLabels`.

Also, I would split this function into 2, one that deals with labels and second one with deprecated labels.

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -565,94 +565,69 @@ func (h *nmcReconcilerHelperImpl) UpdateNodeLabelsAndRecordEvents(ctx context.Co
-	for _, module := range nmc.Spec.Modules {
-		specLabels[types.NamespacedName{Namespace: module.Namespace, Name: module.Name}] = module.Config
-	}
+	getSpecLabelsAndTheirConfigs(specLabels, nmc)
```
This function should return a map.

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -565,94 +565,69 @@ func (h *nmcReconcilerHelperImpl) UpdateNodeLabelsAndRecordEvents(ctx context.Co
-			"Module %s unloaded from the kernel",
-			nsn.String(),
-		)
+	h.recordModuleEvents(&node, unloaded, "unload")
+	h.recordModuleEvents(&node, loaded, "load")
```
@yevgeny-shnaidman @TomerNewman How about we extract the events stuff to the caller of `UpdateNodeLabelsAndRecordEvents` instead of having 1 function doing "multiple different things?

ybettan commented:
> It looks good to me , but i have another suggestion:
most of the code of UpdateNodeLabelsAndRecordEvents function deals with preparing the labels for the node. What if we define another helper interface, that will handle that preparation? This way we can unit-test the UpdateNodeLabelsAndRecordEvents properly and in addition test the labels preparation thoroughly also, since it will be in a different interface. @ybettan , @TomerNewman , what do you think? Or is it an overkill?

I think that the parts that add/remove labels from the nodes should be part of the new `node` interface as long as the label is passed as a parameter.

For other cases, I definitely think we should create another interface and I would also somehow take into consideration `internal->utils->kmmlabels.go` as it is serving a similar purpose.

yevgeny-shnaidman commented:
> > It looks good to me , but i have another suggestion:
> > most of the code of UpdateNodeLabelsAndRecordEvents function deals with preparing the labels for the node. What if we define another helper interface, that will handle that preparation? This way we can unit-test the UpdateNodeLabelsAndRecordEvents properly and in addition test the labels preparation thoroughly also, since it will be in a different interface. @ybettan , @TomerNewman , what do you think? Or is it an overkill?
> 
> I think that the parts that add/remove labels from the nodes should be part of the new `node` interface as long as the label is passed as a parameter.
> 
> For other cases, I definitely think we should create another interface and I would also somehow take into consideration `internal->utils->kmmlabels.go` as it is serving a similar purpose.

I thought about kmmlabels, but its main purpose is to define the labels, to to manipulate what and when should go to wihat node

TomerNewman commented:
Following our slack conversation, I've also split `UpdateNodeLabelsAndRecordEvents` and moved `UpdateNodeLabels` to Node package.
I've also added and changed the unit test accordingly.
I think the PR is way to big, what do we think about removing some of the features into new PR\s? @ybettan @yevgeny-shnaidman

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+	GetSpecLabelsAndTheirConfigs(specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig, nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+	GetStatusLabelsAndTheirConfigs(statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig, nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+}
+type labelPreparationHelperImpl struct {
+}
```
```suggestion
type labelPreparationHelperImpl struct {}
```

ybettan commented on internal/node/node.go:
```diff
@@ -72,3 +77,40 @@ func (n *node) FindNodeCondition(cond []v1.NodeCondition, conditionType v1.NodeC
 	return nil
 }
+
+func (n *node) AddModulesLabelsToNode(modules []types.NamespacedName, node *v1.Node) {
```
Who is using this method?

TomerNewman replied:
my bad, should have deleted it sooner, its fixed now

ybettan commented on internal/node/node.go:
```diff
@@ -13,6 +17,7 @@ type Node interface {
 	GetNodesListBySelector(ctx context.Context, selector map[string]string) ([]v1.Node, error)
 	GetNumTargetedNodes(ctx context.Context, selector map[string]string) (int, error)
 	FindNodeCondition(cond []v1.NodeCondition, conditionType v1.NodeConditionType) *v1.NodeCondition
+	UpdateNodeLabels(ctx context.Context, deprecatedNodeModuleReadyLabels sets.Set[string], loaded []types.NamespacedName, unloaded []types.NamespacedName, node *v1.Node) error
```
```suggestion
	UpdateLabels(ctx context.Context, node *v1.Node, toBeAdded, toBeRemoved []string) error
```

Few comments here:
1. Everything in this package is related to nodes, therefore, there the name `UpdateNodeLabels` can just be `UpdateLabels`.
2. The main object here is the node, therefore, it should be the first param in the method
3. This methods know nothing about `modules`, `readyLabels` etc - it should be simple - get a slice of labels to be added and a list of labels to be removed and just add/remove them from the node
4. Since you are having an API call to the kubeAPI for each label (`meta.RemoveLabel` one by one), I would split this method into 2 `AddLabels` and `RemoveLabels`. If you do a single API call to the kubeAPI for adding/removing all the laels in a single call then it make sense to have a single method for it.

TomerNewman replied:
understood, this is fixed.
now before `UpdateLabels` is being called, we create those two string slices out of the `type.NamespacedName` slices and the set

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-	if err := h.client.Get(ctx, types.NamespacedName{Name: nmc.Name}, &node); err != nil {
-		return fmt.Errorf("could not get node %s: %v", nmc.Name, err)
-	}
+type LabelPreparationHelper interface {
+	GetDeprecatedKernelModuleReadyLabels(deprecatedNodeModuleReadyLabels sets.Set[string], node v1.Node) sets.Set[string]
+	GetKernelModuleReadyLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName], node v1.Node) sets.Set[types.NamespacedName]
+	GetSpecLabelsAndTheirConfigs(specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig, nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+	GetStatusLabelsAndTheirConfigs(statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig, nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+}
```
Let's make it a non-exported interface (same as `nmcReconcilerHelper`)?
@yevgeny-shnaidman Although all those methods are labels related, won't it be cleaner to define them inside `nmcReconcilerHelper`?

TomerNewman replied:
done, `LabelPreparationHelper` was changed to `labelPreparationHelper`

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-	// get all the kernel module ready labels of the node
-	nodeModuleReadyLabels := sets.New[types.NamespacedName]()
-	deprecatedNodeModuleReadyLabels := sets.New[string]()
+func NewLabelPreparationHelper() LabelPreparationHelper {
```
If we don't converge all those methods into `nmcReconcilerHelper` then let's make this function private (same as `newNmcReconcilerHelper`

TomerNewman replied:
we use `NewLabelPreparationHelper` in cmd/manager/main.go line 124.
so I think that if we use this interface, we need this function to be Public

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-	if err := h.client.Get(ctx, types.NamespacedName{Name: nmc.Name}, &node); err != nil {
-		return fmt.Errorf("could not get node %s: %v", nmc.Name, err)
-	}
+type LabelPreparationHelper interface {
```
Are all those methods has to be exported? Even if the interface isn't exported?

TomerNewman replied:
so all of the methods in labelPreparationHelper should be private?
I was trying to take an example from `podManager` interface

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
 			nodeModuleReadyLabels.Insert(types.NamespacedName{Namespace: namespace, Name: name})
 		}
+	}
+	return nodeModuleReadyLabels
```
A `set` is a map (pointer) which means that when you `Insert` to it, it get updated, therefore there is not need to also return it.

I would define the Set inside the function and return it.
A function that has no state is easier to debug and preferable IMO

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+	return &labelPreparationHelperImpl{}
+}
 
+func (lphi *labelPreparationHelperImpl) GetKernelModuleReadyLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName], node v1.Node) sets.Set[types.NamespacedName] {
```
```suggestion
func (lphi *labelPreparationHelperImpl) GetNodeKernelModuleReadyLabels(node v1.Node) sets.Set[types.NamespacedName] {
```

ybettan replied:
Same for other methods in this interface.

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+}
 
-	return nil
+func (h *nmcReconcilerHelperImpl) removeOrphanedLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
```
Any reason this method isn't part of the help interface as well?

TomerNewman replied:
I thought because its an helper (inner) function, so if someone wants to implement this interface outside of this package, we wont have to force him to implement inner functions

ybettan replied:
You main point is correct, but we can say it about all the methods in the helper function. No one else will implement that interface and we only do it for easier unit-testing so having those methods in the interface is a better option IMO.

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+}
 
-	for _, nsn := range unloaded {
+func (h *nmcReconcilerHelperImpl) RecordEvents(loaded []types.NamespacedName, unloaded []types.NamespacedName, node *v1.Node) {
```
Can we use `loadedModules` and `unloadedModules`?

TomerNewman replied:
yes.
also, do you think `loadedModules` and `unloadedModules` should be []string instead?

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+func (h *nmcReconcilerHelperImpl) removeOrphanedLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+	specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+	unloaded []types.NamespacedName) []types.NamespacedName {
```
Can we define this one inside the function?

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+func (h *nmcReconcilerHelperImpl) addEqualLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+	specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+	loaded []types.NamespacedName) []types.NamespacedName {
```
Can we define this one inside the method?

TomerNewman replied:
done

yevgeny-shnaidman commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,138 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+type labelPreparationHelperImpl struct{}
 
-	// get all the kernel module ready labels of the node
+func NewLabelPreparationHelper() labelPreparationHelper {
```
We don't need this function to be public. This is an inner interface to the package, it can be initialized by the NewNMCReconcile , just like all other helper interfaces

ybettan replied:
I agree, but you will need to manually create the object when testing in order to use a mock since there is no dependency injection anymore.

yevgeny-shnaidman replied:
we do it all the time in the unit-test

TomerNewman replied:
done.
now it is initialized in `newNMCReconcilerHelper` instead of `NewNMCReconcile` because we need it in  `nmcReconcilerHelperImpl` struct

ybettan commented on internal/utils/convert_types.go:
```diff
@@ -0,0 +1,26 @@
+	return convertedSlice
+}
+
+func SetToSlice[T comparable](set sets.Set[T]) []T {
+
+	slice := make([]T, 0, set.Len())
+
+	for item := range set {
+		slice = append(slice, item)
+	}
+	return slice
+}
```
This is already implemented in the `sets` package as `set.UnsortedList`, therefore, we can remove this function.

TomerNewman replied:
File was deleted

ybettan commented on internal/utils/convert_types.go:
```diff
@@ -0,0 +1,26 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+func ConvertNamespacedNameToStrings(toBeConverted []types.NamespacedName) []string {
```
This function isn't converting `[]NamespacedName` to `[]string` but `[]NamespacedName` to a very specific KMM label list.

How about we just return lists from `removeOrphanedLabels` and `addEqualLabels` instead of lists of `NamespacedName` and then write another logic to convert it?

We can completely remove this file IMO.

TomerNewman replied:
@ybettan so in `removeOrphanedLabels` and `addEqualLabels` we will return []string and we should have a function that will convert the []string label list to a []type.NamespaceName list?

ybettan replied:
No. I mean instead of:
* Returning `[]NamespacedName` and then write code to convert it

Let's just:
* Return `[]string` (then we don't need any logic to convert it) - the only use we do with `[]NamespacedName` is to convert it so it is just redundant.

TomerNewman replied:
The file was deleted

ybettan commented on internal/node/node.go:
```diff
@@ -72,3 +74,34 @@ func (n *node) FindNodeCondition(cond []v1.NodeCondition, conditionType v1.NodeC
+	return nil
+}
+
+func (n *node) addLabels(node *v1.Node, toBeAdded []string) {
```
```suggestion
func (n *node) addLabels(node *v1.Node, labels []string) {
```

TomerNewman replied:
done

ybettan commented on internal/node/node.go:
```diff
@@ -72,3 +74,34 @@ func (n *node) FindNodeCondition(cond []v1.NodeCondition, conditionType v1.NodeC
+	}
+}
+
+func (n *node) removeLabels(node *v1.Node, toBeRemoved []string) {
```
```suggestion
func (n *node) removeLabels(node *v1.Node, labels []string) {
```

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -213,22 +220,25 @@ type nmcReconcilerHelper interface {
 	RemovePodFinalizers(ctx context.Context, nodeName string) error
 	SyncStatus(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig) error
-	UpdateNodeLabelsAndRecordEvents(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig) error
+	PrepareAndUpdateNodeLabels(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig, node *v1.Node) ([]types.NamespacedName, []types.NamespacedName, error)
```
```suggestion
	UpdateNodeLabels(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig, node *v1.Node) ([]types.NamespacedName, []types.NamespacedName, error)
```

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,138 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-	return nil
+func (h *nmcReconcilerHelperImpl) removeOrphanedLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+	specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName {
```
```suggestion
	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []string {
```
Can we just return a label list instead? No reason to build an object just to parse it later on.

TomerNewman replied:
following our meeting, I left it as it was

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,138 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+}
+func (h *nmcReconcilerHelperImpl) addEqualLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+	specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName {
```
```suggestion
	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []string {
```

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +565,138 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+	}
+	return unloaded
+}
+func (h *nmcReconcilerHelperImpl) addEqualLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
```
I would make this function an interface method as well.

ybettan commented on internal/node/node.go:
```diff
@@ -72,3 +76,45 @@ func (n *node) FindNodeCondition(cond []v1.NodeCondition, conditionType v1.NodeC
+
+type nodeHelperImpl struct{}
+
+type nodeHelper interface {
+	addLabels(node *v1.Node, labels []string)
+	removeLabels(node *v1.Node, labels []string)
+}
```
I am not sure a helper interface is the best option here. I would make them non-exported functions (not methods) in the package.

Here are some points to consider when deciding if something should be an interface/function.
1. As a general rule of thumb, if a method isn't using its "self" object then it should be a function.
2. If an API is exported, using its "self" object and wish to be mocked in unit-tests then we use an exported interface.
3. In some cases, we are having exceptions in the form of "helper, non-exported interfaces" - in cases where a packages has a lot of non-exported functions/methods we sometimes choose to put them inside an helper interface (even if the "self" object isn't being used just for the ease of unit-testing the package because it becomes difficult to test.

In this case, `addLabels` and `removeLabels`:
1. Does not use the "self" (node) object - leaning towards being functions rather than methods.
2. Are pretty short/simple and there are only 2 "helper" functions so it is not too difficult to unit-test `UpdateLabels`

Therefore I would make those non-exported functions and remove the helper interface in this package.

TomerNewman replied:
@ybettan  so in that case, you think that `addLabels` and `removeLabels` should not be mocked? 
they are pretty simple, right, but mocking them helps for example in line 173 in node_test.go

TomerNewman replied:
deleted the `nodeHelper` interface

ybettan replied:
> so in that case, you think that addLabels and removeLabels should not be mocked?

Indeed. In this case I don't think the mocking benefit worth writing another interface.

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -213,14 +219,16 @@ type nmcReconcilerHelper interface {
 	SyncStatus(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig) error
-	UpdateNodeLabelsAndRecordEvents(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig) error
+	UpdateNodeLabels(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig, node *v1.Node) ([]types.NamespacedName, []types.NamespacedName, error)
+	RecordEvents(loadedModules []types.NamespacedName, unloadedModules []types.NamespacedName, node *v1.Node)
```
```suggestion
	RecordEvents(node *v1.Node, loadedModules, unloadedModules []types.NamespacedName, )
```

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +564,154 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+	getNodeKernelModuleReadyLabels(node v1.Node) sets.Set[types.NamespacedName]
+	getSpecLabelsAndTheirConfigs(nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+	getStatusLabelsAndTheirConfigs(nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+	addEqualLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+		specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+		statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName
```
```suggestion
	addEqualLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
		specLabels, statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName
```

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +564,154 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+	addEqualLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+		specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+		statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName
+	removeOrphanedLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+		specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+		statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName
```
```suggestion
	removeOrphanedLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
		specLabels, statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName
```

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +564,154 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
 
-			loaded = append(loaded, nsn)
-		}
+	unloadedLabels := []string{}
+	loadedLabels := []string{}
```
```suggestion
        var loadedLabels unloadedLabels []string
```

See the difference:
https://go.dev/play/p/JBrR0homl_7

Using `nil` will result in an error in case we do something illegal with the slices (which is better than having a bug that isn't erroring).

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +564,154 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-	if err := h.client.Patch(ctx, &node, patchFrom); err != nil {
-		return fmt.Errorf("could not patch node: %v", err)
+	if err := h.nodeAPI.UpdateLabels(ctx, node, loadedLabels, unloadedLabels); err != nil {
+		return nil, nil, fmt.Errorf("could not patch node: %v", err)
```
```suggestion
		return nil, nil, fmt.Errorf("could not update labels on the node: %v", err)
```

TomerNewman replied:
done

ybettan commented on internal/controllers/nmc_reconciler.go:
```diff
@@ -555,104 +564,149 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
 
-			loaded = append(loaded, nsn)
-		}
+	var loadedLabels, unloadedLabels []string
+
+	for label := range deprecatedNodeModuleReadyLabels {
+		unloadedLabels = append(unloadedLabels, label)
+	}
```
```suggestion
var loadedLabels []string
unloadedLabels := deprecatedNodeModuleReadyLabels.UnsortedList()

for _, label := range nsnLabelsToBeRemoved {
    unloadedLabels = append(unloadedLabels, utils.GetKernelModuleReadyNodeLabel(label.Namespace, label.Name))
}
```

ybettan commented on internal/node/node.go:
```diff
@@ -13,6 +14,9 @@ type Node interface {
 	GetNumTargetedNodes(ctx context.Context, selector map[string]string) (int, error)
 	FindNodeCondition(cond []v1.NodeCondition, conditionType v1.NodeConditionType) *v1.NodeCondition
+	UpdateLabels(ctx context.Context, node *v1.Node, toBeAdded, toBeRemoved []string) error
+	addLabels(node *v1.Node, labels []string)
+	removeLabels(node *v1.Node, labels []string)
```
Let's remove them from the interface and make then non-exported *functions*.

https://github.com/kubernetes-sigs/kernel-module-management/pull/850#discussion_r1673821877

ybettan commented on internal/node/node_test.go:
```diff
@@ -220,3 +269,64 @@ var _ = Describe("FindNodeCondition", func() {
+	It("Should add labels", func() {
+		labels := []string{kernelModuleReadyNodeLabel}
+
+		Expect(node.Labels).ToNot(HaveKey(kernelModuleReadyNodeLabel))
```
```suggestion
```

ybettan commented on internal/node/node_test.go:
```diff
@@ -220,3 +269,64 @@ var _ = Describe("FindNodeCondition", func() {
+		n.addLabels(&node, labels)
+		Expect(node.Labels).To(HaveKey(kernelModuleReadyNodeLabel))
+	})
+	It("Should not add labels", func() {
+		var labels []string
+
+		Expect(node.Labels).ToNot(HaveKey(kernelModuleReadyNodeLabel))
+		n.addLabels(&node, labels)
+		Expect(node.Labels).ToNot(HaveKey(kernelModuleReadyNodeLabel))
+	})
+})
```
We can remove this test - it is not really testing anything different than the previous spec.

TomerNewman replied:
@ybettan  why not? in theory `kernelModuleReadyNodeLabel` can be added no matter what in `addLabels`, so dont we need to test in both directions? (one with kernelModuleReadyNodeLabel and one without kernelModuleReadyNodeLabel)

ybettan replied:
In the previous test, you start with an empty node, add a label and check if the label was added.
In this test, you start with an empty node, call `addLabels` with an empty list and check that nothing happened.
I am not sure what added value this test have over the previous one.

TomerNewman replied:
test is removed

ybettan commented on internal/node/node_test.go:
```diff
@@ -220,3 +269,64 @@ var _ = Describe("FindNodeCondition", func() {
+	It("Should remove labels", func() {
+		labels := []string{kernelModuleReadyNodeLabel}
+
+		Expect(node.Labels).To(HaveKey(kernelModuleReadyNodeLabel))
```
```suggestion
```

ybettan commented on internal/node/node_test.go:
```diff
@@ -220,3 +269,64 @@ var _ = Describe("FindNodeCondition", func() {
+		n.removeLabels(&node, labels)
+		Expect(node.Labels).ToNot(HaveKey(kernelModuleReadyNodeLabel))
+	})
+	It("Should not remove labels", func() {
+		var labels []string
+
+		Expect(node.Labels).To(HaveKey(kernelModuleReadyNodeLabel))
+		n.removeLabels(&node, labels)
+		Expect(node.Labels).To(HaveKey(kernelModuleReadyNodeLabel))
+	})
```
We can remove this test - it is not really testing anything different than the previous spec.

ybettan replied:
Same as https://github.com/kubernetes-sigs/kernel-module-management/pull/850#discussion_r1677109887.

It is not crucial at all to have it. If you think there is benefit in it then feel free to keep it. I don't see a lot of value in it but it is your decision :)

TomerNewman replied:
test is removed

ybettan commented on internal/controllers/nmc_reconciler_test.go:
```diff
@@ -143,6 +158,9 @@ var _ = Describe("NodeModulesConfigReconciler_Reconcile", func() {
 			reflect.TypeOf((*context.Context)(nil)).Elem(),
 		)
 
+		kubeClient.EXPECT().Get(ctx, types.NamespacedName{Name: nmc.Name}, &node).Return(err)
```
```suggestion
		kubeClient.EXPECT().Get(ctx, types.NamespacedName{Name: nmc.Name}, &node).Return(nil)
```

If we don't return any error, then let's return `nil` it is clearer then searching what `err` got as its value.

Also, any reason we don't put this line inside the `InOrder` block?

TomerNewman replied:
fixed, moved it also to the InOrder section

ybettan commented on internal/controllers/nmc_reconciler_test.go:
```diff
@@ -143,6 +158,9 @@ var _ = Describe("NodeModulesConfigReconciler_Reconcile", func() {
 		)
 
+		kubeClient.EXPECT().Get(ctx, types.NamespacedName{Name: nmc.Name}, &node).Return(err)
+		Expect(err).NotTo(HaveOccurred())
```
What are we testing here? If `err` was not initialized, then it will be `nil` by default.

TomerNewman replied:
fixed, moved `kubeClient.EXPECT().Get(ctx, types.NamespacedName{Name: nmc.Name}, &node).Return(err)` to the InOrder section (`Expect(err).NotTo(HaveOccurred())` should not be there indeed)

ybettan commented on internal/controllers/nmc_reconciler_test.go:
```diff
@@ -205,6 +229,8 @@ var _ = Describe("NodeModulesConfigReconciler_Reconcile", func() {
 		contextWithValueMatch := gomock.AssignableToTypeOf(
 			reflect.TypeOf((*context.Context)(nil)).Elem(),
 		)
+		kubeClient.EXPECT().Get(ctx, types.NamespacedName{Name: nmc.Name}, &node).Return(err)
+		Expect(err).NotTo(HaveOccurred())
```
Same as https://github.com/kubernetes-sigs/kernel-module-management/pull/850/files#r1677111464

TomerNewman replied:
fixed, moved `kubeClient.EXPECT().Get(ctx, types.NamespacedName{Name: nmc.Name}, &node).Return(err)` to the InOrder section (`Expect(err).NotTo(HaveOccurred())` should not be there indeed)

ybettan commented on internal/controllers/nmc_reconciler_test.go:
```diff
@@ -217,10 +243,10 @@ var _ = Describe("NodeModulesConfigReconciler_Reconcile", func() {
 
-		_, err := r.Reconcile(ctx, req)
+		_, err = r.Reconcile(ctx, req)
 		Expect(err).ToNot(BeNil())
```
This comment is for purpose learning only as it is not directly related to your PR.

This is not a great validation. The description of the spec is `should complete all the reconcile functions and return combined error`, therefore, the validation should be to (first return different error messages and then) validate that the combined error contain each one of the error messages of the individual errors.

ybettan replied:
You don't have to add it to the PR so we can merge it faster. You can fix it in a future PR if you like. But to answer your question, the answer is YES.

One way to check it is like this example:
https://github.com/kubernetes-sigs/kernel-module-management/blob/c44a5e8d908a3bdd039f1f5cb8ee18663499eb6e/internal/registry/registry_test.go#L57

ybettan commented on internal/controllers/nmc_reconciler_test.go:
```diff
@@ -1158,39 +1184,143 @@ var _ = Describe("nmcReconcilerHelperImpl_RemovePodFinalizers", func() {
+				).Return(nil),
+		)
+		_, _, err := wh.UpdateNodeLabels(ctx, &nmc, &node)
+		Expect(err).ToNot(HaveOccurred())
```
This is only testing that no error was returned. What about the part that validates that the node was updated correctly?

TomerNewman replied:
I added another Except for the new node label (next line)

ybettan commented on internal/node/node_test.go:
```diff
@@ -118,6 +120,59 @@ var _ = Describe("GetNodesListBySelector", func() {
+		Expect(node.Labels).To(HaveKey(loadedKernelModuleReadyNodeLabel))
+	})
+
+	It("Should fail to patch node", func() {
```
```suggestion
	It("Should fail if we fail to patch node", func() {
```

ybettan commented:
/approve
/assign @yevgeny-shnaidman

yevgeny-shnaidman commented:
/lgtm
